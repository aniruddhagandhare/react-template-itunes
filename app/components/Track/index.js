/**
 *
 * Track
 *
 */

import React, { memo, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card, Typography } from 'antd';
import { PauseOutlined, CaretRightFilled } from '@ant-design/icons';
import { colors, media } from '@app/themes';

import { FormattedMessage as T } from 'react-intl';

const CustomCard = styled.div`
  display: grid;
  ${media.mobile.min(`
     grid-template-columns: 10em 1fr;
   `)}
  gap: 1.4em;
  img {
    width: 100%;
    height: 100%;
    object-fit: cover;
    border-radius: 5px;
  }
  div.inner-container {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    audio {
      width: 100%;
    }
  }
`;
const AudioButton = styled.div`
  height: 3.5em;
  width: 3.5em;
  border-radius: 3.5em;
  background-color: ${colors.secondary};
  color: ${colors.primary};
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  svg {
    height: 1.4em;
    width: 1.4em;
  }
`;

const AudioContainer = styled.div`
  display: flex;
  align-items: center;
`;

const Slider = styled.div`
  margin-left: 1em;
  border-radius: 1em;
  height: 0.3em;
  background: ${colors.primary};
`;
const { Text } = Typography;

function Track({ track }) {
  const audioRef = useRef(null);
  const [playing, setPlaying] = useState(false);
  const currentTimeRef = useRef(null);

  const togglePlayState = () => {
    if (!playing) {
      setPlaying(!playing);
      audioRef.current.play();
    } else {
      setPlaying(!playing);
      audioRef.current.pause();
    }
  };
  audioRef &&
    audioRef.current &&
    audioRef.current.addEventListener(
      'timeupdate',
      e => (currentTimeRef.current.style.width = (audioRef.current.currentTime / audioRef.current.duration) * 100 + '%')
    );
  return (
    <Card
      title={
        <div>
          <T id="track_name" values={{ trackName: track.trackName }} /> by{' '}
          <a
            target="_blank"
            style={{ color: track.artistViewUrl ? colors.primary : colors.text }}
            href={track.artistViewUrl}
            rel="noreferrer"
          >
            <T id="artist_name" values={{ artistName: track.artistName }} />
          </a>
        </div>
      }
      extra={
        <a style={{ color: colors.primary }} target="_blank" href={track.trackViewUrl} rel="noreferrer">
          <T id="view_original" />
        </a>
      }
    >
      <CustomCard className="custom">
        <img src={track.artworkUrl100} />
        <div className="inner-container">
          <Text>
            <T id="track_heading" />
            {' - '}
            <strong>
              {track.collectionName ? (
                <T id="collection_name" values={{ collectionName: track.collectionName }} />
              ) : (
                <T id="track_heading_not_found" />
              )}
            </strong>
          </Text>
          <br></br>
          <div>
            <audio ref={audioRef} preload="auto">
              <source src={track.previewUrl}></source>
            </audio>
            <AudioContainer className="button-container">
              <AudioButton onClick={togglePlayState}>{playing ? <PauseOutlined /> : <CaretRightFilled />}</AudioButton>
              <Slider ref={currentTimeRef}></Slider>
            </AudioContainer>
          </div>
        </div>
      </CustomCard>
    </Card>
  );
}

Track.propTypes = {
  track: PropTypes.object
};

export default memo(Track);
