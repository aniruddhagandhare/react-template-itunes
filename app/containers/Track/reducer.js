/*
 *
 * Track reducer
 *
 */
import produce from 'immer';
import { createActions } from 'reduxsauce';

export const initialState = {
  track: null,
  error: null
};

export const { Types: trackTypes, Creators: trackCreators } = createActions({
  getTrackById: ['trackId'],
  successGetTrackById: ['track'],
  errorGetTrackById: ['error']
});

/* eslint-disable default-case, no-param-reassign */
export const trackReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case trackTypes.GET_TRACK_BY_ID:
        draft.track = null;
        break;
      case trackTypes.SUCCESS_GET_TRACK_BY_ID:
        draft.track = action.track.results[0];
        break;
      case trackTypes.ERROR_GET_TRACK_BY_ID:
        console.log(action);
        draft.error = get(action.error, 'message', 'Something went wrong');
        break;
    }
  });

export default trackReducer;
