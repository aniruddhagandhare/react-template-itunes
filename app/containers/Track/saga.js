import { takeLatest, put, call } from 'redux-saga/effects'
import { getTrack } from '@services/itunesApi';
import { trackTypes, trackCreators } from './reducer'

// Individual exports for testing
const { GET_TRACK_BY_ID } = trackTypes;
const { successGetTrackById, errorGetTrackById} = trackCreators;
export function *defaultFunction (action) {
  const response = yield call(getTrack, action.trackId)
  const { ok, data } = response;
  if(ok) {
    yield put(successGetTrackById(data))
  }
  else {
    yield put(errorGetTrackById(data))
  }
}

export default function* trackSaga() {
  yield takeLatest(GET_TRACK_BY_ID, defaultFunction)
}