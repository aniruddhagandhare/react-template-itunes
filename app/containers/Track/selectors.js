import { get } from 'lodash'
import { createSelector } from 'reselect'
import { initialState } from './reducer'

/**
 * Direct selector to the track state domain
 */

const selectTrackDomain = state => {
  return state.trackContainer || initialState
}

const makeSelectTrack = () =>
  createSelector(selectTrackDomain, substate => substate)

export const selectTrack = () => createSelector(selectTrackDomain, substate => get(substate, 'track', {}))

export default makeSelectTrack
export { selectTrackDomain }
