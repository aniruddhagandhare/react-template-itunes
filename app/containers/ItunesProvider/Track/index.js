/**
 *
 * Track
 *
 */
import React, { memo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl } from 'react-intl';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Skeleton, Alert } from 'antd';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import { useInjectSaga } from '@utils/injectSaga';
import T from '@app/components/T';
import If from '@app/components/If';
import IndividualTrack from '@app/components/IndividualTrack';
import { fonts, colors } from '@app/themes';
import PropTypeContants from '@app/utils/PropTypeContants';
import { selectError, selectLoading, selectTrack } from '../selectors';
import saga from '../saga';
import { itunesCreators } from '../reducer';

const TrackWrapper = styled.div`
  max-width: 50em;
  margin: 0 auto;
  padding: 1.3em;
  position: relative;
`;
const CustomLink = styled(Link)`
  && {
    ${fonts.size.small()};
    color: ${colors.primary};
    display: block;
    margin-bottom: ${props => (props.mb ? `${props.mb}em` : 0)};
  }
`;
export function Track({ track, dispatchGetTrackById, match, error, loading }) {
  useInjectSaga({ key: 'track', saga });
  const { params } = match;
  const trackId = params.trackId;
  useEffect(() => {
    dispatchGetTrackById(trackId);
  }, []);
  return (
    <TrackWrapper>
      <CustomLink to="/" mb="2">
        <T id="back_to_all_tracks" />
      </CustomLink>
      <If condition={track} otherwise={<Skeleton loading={loading} active />}>
        <IndividualTrack track={track} showInfo={true} />
      </If>
      <If condition={error}>
        <Alert message={<T id="track_error" values={{ error }} />} type="error" />
      </If>
    </TrackWrapper>
  );
}

Track.propTypes = {
  track: PropTypes.shape(PropTypeContants).isRequired,
  dispatchGetTrackById: PropTypes.func,
  match: PropTypes.object,
  error: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  loading: PropTypes.bool
};

const mapStateToProps = createStructuredSelector({
  track: selectTrack(),
  error: selectError(),
  loading: selectLoading()
});

function mapDispatchToProps(dispatch) {
  const { requestGetTrackById } = itunesCreators;
  return {
    dispatchGetTrackById: trackId => dispatch(requestGetTrackById(trackId))
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default compose(
  withConnect,
  memo
)(Track);

export const TrackTest = compose(injectIntl)(Track);
