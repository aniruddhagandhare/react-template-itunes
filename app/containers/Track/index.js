/**
 *
 * Track
 *
 */

import React, { memo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl, FormattedMessage as T } from 'react-intl';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import { Skeleton } from 'antd';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import { useInjectSaga } from '@utils/injectSaga';
import { selectTrack } from './selectors';
import { trackCreators } from './reducer';
import saga from './saga';
import { colors } from '@app/themes/index';
import { fonts } from '@app/themes';
import If from '@app/components/If';
import IndividualTrack from '@app/components/IndividualTrack';

const TrackWrapper = styled.div`
  max-width: 80em;
  margin: 0 auto;
  padding: 1.3em;
  position: relative;
`;
const CustomLink = styled(Link)`
  && {
    ${fonts.size.small()};
    color: ${colors.primary};
    display: block;
    margin-bottom: ${props => (props.mb ? `${props.mb}em` : 0)};
  }
`;
export function Track({ track, dispatchGetTrackById, match }) {
  useInjectSaga({ key: 'track', saga });
  const { params } = match;
  const trackId = params.trackId;
  useEffect(() => {
    dispatchGetTrackById(trackId);
  }, []);
  return (
    <TrackWrapper>
      <CustomLink to="/demo" mb="2">
        <T id="back_to_all_tracks" />
      </CustomLink>
      <If condition={track} otherwise={<Skeleton active />}>
        <IndividualTrack track={track} showInfo={true} />
      </If>
    </TrackWrapper>
  );
}

Track.propTypes = {
  track: PropTypes.object,
  dispatchGetTrackById: PropTypes.func,
  match: PropTypes.object
};

const mapStateToProps = createStructuredSelector({
  track: selectTrack()
});

function mapDispatchToProps(dispatch) {
  const { requestGetTrackById } = trackCreators;
  return {
    dispatchGetTrackById: trackId => dispatch(requestGetTrackById(trackId))
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps
);

export default compose(
  withConnect,
  memo
)(Track);

export const TrackTest = compose(injectIntl)(Track);
