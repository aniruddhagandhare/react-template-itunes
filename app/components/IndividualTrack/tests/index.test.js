/**
 *
 * Tests for Track
 *
 */

import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import { renderWithIntl } from '@utils/testUtils';
import Track from '../index';

describe('<Track />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<Track />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 Track component', () => {
    const { getAllByTestId } = renderWithIntl(<Track />);
    expect(getAllByTestId('track').length).toBe(1);
  });
});
