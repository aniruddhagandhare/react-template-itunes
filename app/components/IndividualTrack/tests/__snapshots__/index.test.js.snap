// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Track /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ant-card ant-card-bordered IndividualTrack__CustomCard-sc-1gi6ox1-4 eMahDM"
      data-testid="track"
    >
      <div
        class="ant-card-body"
      >
        <div
          class="IndividualTrack__Content-sc-1gi6ox1-0 hkHlBc"
        >
          <div
            class="IndividualTrack__AudioContainer-sc-1gi6ox1-3 jxLaGq"
          >
            <audio
              preload="auto"
            >
              <source
                src="https://audio-ssl.itunes.apple.com/itunes-assets/Music/v4/de/1d/c1/de1dc139-7898-5fbb-da66-06ce88426178/mzaf_1897766559436846339.plus.aac.p.m4a"
              />
            </audio>
            <div
              class="IndividualTrack__AudioButton-sc-1gi6ox1-2 ligNUl"
            >
              <span
                aria-label="caret-right"
                class="anticon anticon-caret-right"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="caret-right"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 1024 1024"
                  width="1em"
                >
                  <path
                    d="M715.8 493.5L335 165.1c-14.2-12.2-35-1.2-35 18.5v656.8c0 19.7 20.8 30.7 35 18.5l380.8-328.4c10.9-9.4 10.9-27.6 0-37z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            class="IndividualTrack__RightContent-sc-1gi6ox1-1 gPpiCv"
          >
            <p
              class="T__StyledText-gjlic1-0 hCAeFa IndividualTrack__StyledT-sc-1gi6ox1-5 imKGmW"
              data-testid="t"
            >
              Just Another Day
            </p>
            <a
              rel="noreferrer"
              style="color: rgb(223, 94, 94);"
              target="_blank"
            >
              <p
                class="T__StyledText-gjlic1-0 bVNisQ IndividualTrack__StyledT-sc-1gi6ox1-5 imKGmW"
                data-testid="t"
              >
                by Drake
              </p>
            </a>
            <div
              class="IndividualTrack__VSpace-sc-1gi6ox1-6 cjhaPD"
            />
            <p
              class="T__StyledText-gjlic1-0 cbAvsY IndividualTrack__StyledT-sc-1gi6ox1-5 imKGmW"
              data-testid="t"
            >
              Price - $. 0.99
            </p>
            <span
              class="ant-tag ant-tag-has-color"
              style="background-color: rgb(223, 94, 94);"
            >
              <p
                class="T__StyledText-gjlic1-0 bVNisQ IndividualTrack__StyledT-sc-1gi6ox1-5 imKGmW"
                data-testid="t"
              >
                explicit
              </p>
            </span>
            <span
              class="ant-tag ant-tag-has-color"
              style="background-color: rgb(223, 94, 94);"
            >
              <p
                class="T__StyledText-gjlic1-0 bVNisQ IndividualTrack__StyledT-sc-1gi6ox1-5 imKGmW"
                data-testid="t"
              >
                
              </p>
            </span>
          </div>
          <a
            class="IndividualTrack__LinkTo-sc-1gi6ox1-7 dfSsTA"
            target="_blank"
          >
            <p
              class="T__StyledText-gjlic1-0 bVNisQ IndividualTrack__StyledT-sc-1gi6ox1-5 imKGmW"
              data-testid="t"
            >
              View Track
            </p>
          </a>
        </div>
        <div
          class="IndividualTrack__SongProgress-sc-1gi6ox1-8 cPDWaK"
        />
      </div>
    </div>
  </div>
</body>
`;
